src/main/java/org/humancellatlas/ingest/query/QueryBuilder.java

```diff
@@ -0,0 +1,66 @@
+package org.humancellatlas.ingest.query;
+
+import org.springframework.data.mongodb.InvalidMongoDbApiUsageException;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class QueryBuilder {
+    public Query build(List<MetadataCriteria> criteriaList, Boolean andCriteria){
+        Query query = new Query();
+
+        List<Criteria> criterias = new ArrayList<>();
+
+        for (MetadataCriteria metadataCriteria : criteriaList) {
+            String field = metadataCriteria.getField();
+            Criteria criteria = Criteria.where(field);
+            switch (metadataCriteria.getOperator()) {
+                case IS:
+                    try {
+                        criteria = criteria.is(metadataCriteria.getValue());
+                    } catch (InvalidMongoDbApiUsageException e) {
+                        throw new IllegalArgumentException(e.getMessage(), e);
+                    }
+                    break;
+                case NE:
+                    criteria = criteria.ne(metadataCriteria.getValue());
+                    break;
+                case GT:
+                    criteria = criteria.gt(metadataCriteria.getValue());
+                    break;
+                case GTE:
+                    criteria = criteria.gte(metadataCriteria.getValue());
+                    break;
+                case LT:
+                    criteria = criteria.lt(metadataCriteria.getValue());
+                    break;
+                case LTE:
+                    criteria = criteria.lte(metadataCriteria.getValue());
+                    break;
+                case IN:
+                    criteria = criteria.in((Collection<?>) metadataCriteria.getValue());
+                    break;
+                case NIN:
+                    criteria = criteria.nin((Collection<?>) metadataCriteria.getValue());
+                    break;
+                case REGEX:
+                    criteria = criteria.regex((String) metadataCriteria.getValue(), "i");
+                    break;
+                default:
+                    throw new IllegalArgumentException(String.format("MetadataCriteria %s is not supported.", metadataCriteria.getOperator()));
+            }
```

Comment:
This long switch statement can be refactored into something like `metadataCriteria.map(criteria)`. In the `Operator` enum, the mapping can be written in. Something like:

    enum Operator {
        GT((c, o) -> c.gt(o)),
        LT((c, o) -> c.lt(o)),
        NE((c, o) -> c.ne(o));

        Operator(BiConsumer<Criteria, Object> action) { // can alternatively use BiFunction
            this.action = action;
        }

        void define(Criteria c, Object o) {
            action.accept(c, o);
        }

    }

Reply:
Looks neat! However, I don't think I have time by end of sprint to apply this. I just extracted this from old code because I want to reuse those. I'll look into this next time.